File: bi_node.rb
class BiNode
  attr_accessor :data, :visited, :key, :left_node, :right_node

  def initialize(data:, key: nil, left_node: nil, right_node: nil)
    @data = data
    @key = key
    @visited = false
    @left_node = left_node
    @right_node = right_node
  end

  def get_all_nodes
    return [] if nil?
    result = [self]
    left = @left_node.nil? ? [] : @left_node.get_all_nodes
    right = @right_node.nil? ? [] : @right_node.get_all_nodes
    result.push(*left, *right)
    return result
  end

  def print
    return 'nil' if self.nil?
    result = [@data]
    left, right = @left_node, @right_node
    while !left.nil? || !right.nil? do
      unless right.nil?
        result.push("#{right.data}")
        right = right.right_node
      end
      unless left.nil?
        result.prepend("#{left.data}")
        left = left.left_node
      end
    end
    result = result.join("<->")
    return result + "\n"
  end
end

File: bi_node.py
class BiNode:
  def __init__(self, *, data, left_node=None, right_node=None):
    self.data = data
    self.left_node = left_node
    self.right_node = right_node
    self.visited = False

  def get_all_nodes(self):
    if self is None: return []
    result = [self]
    left = [] if self.left_node is None else self.left_node.get_all_nodes()
    right = [] if self.right_node is None else self.right_node.get_all_nodes()
    result.extend(left)
    result.extend(right)
    return result

def test():
  n = BiNode(
    data=4,
    left_node=BiNode(
      data=5,
      left_node=BiNode(data=8),
      right_node=BiNode(data=9)),
    right_node=BiNode(
      data=6,
      left_node=BiNode(
        data=19,
        left_node=BiNode(data=20),
      )),
  )
  for node in n.get_all_nodes():
    print(node.data)

if __name__ == '__main__':
    test()
